tests/MongoDB.Driver.Core.Tests/Core/Servers/ServerTests.cs

```diff
@@ -266,6 +267,145 @@ public void Dispose_should_dispose_the_server()
+            mockConnectionPool.Setup(p => p.AcquireConnectionAsync(It.IsAny<CancellationToken>())).ReturnsAsync(mockConnection.Object);
+            var mockConnectionPoolFactory = new Mock<IConnectionPoolFactory>();
+            mockConnectionPoolFactory
+                .Setup(f => f.CreateConnectionPool(It.IsAny<ServerId>(), _endPoint))
+                .Returns(mockConnectionPool.Object);
+            var mockMonitorServerDescription = new ServerDescription(serverId, _endPoint);
+            var mockServerMonitor = new Mock<IServerMonitor>();
+            mockServerMonitor.SetupGet(m => m.Description).Returns(mockMonitorServerDescription);
+            void MockMonitorInvalidate(string reason)
+            {
+                var currentDescription = mockServerMonitor.Object.Description;
+                mockServerMonitor.SetupGet(m => m.Description).Returns(currentDescription.With(reason));
+            }
+            mockServerMonitor
+                .Setup(m => m.Invalidate(It.IsAny<string>()))
+                .Callback((string reason) => MockMonitorInvalidate(reason));
+            var mockServerMonitorFactory = new Mock<IServerMonitorFactory>();
+            mockServerMonitorFactory.Setup(f => f.Create(It.IsAny<ServerId>(), _endPoint)).Returns(mockServerMonitor.Object);
+
+            _capturedEvents = new EventCapturer();
+            _settings = new ServerSettings(heartbeatInterval: Timeout.InfiniteTimeSpan);
+            var subject = new Server(_clusterId, _clusterClock, _clusterConnectionMode, _settings, _endPoint, mockConnectionPoolFactory.Object, mockServerMonitorFactory.Object, _capturedEvents);
+            subject.Initialize();
+
+            IChannelHandle channel = null;
+            Exception exception;
+            if (async)
+            {
+                exception = Record.Exception(() => channel = subject.GetChannelAsync(CancellationToken.None).GetAwaiter().GetResult());
+            }
+            else
+            {
+                exception  = Record.Exception(() => channel = subject.GetChannel(CancellationToken.None));
+            }
+
+            channel.Should().BeNull();
+            exception.Should().Be(openConnectionException);
+            subject.Description.Type.Should().Be(ServerType.Unknown);
+            subject.Description.ReasonChanged.Should().Contain("ChannelException during handshake");
+            mockServerMonitor.Verify(m => m.Invalidate(It.IsAny<string>()), Times.Once);
+            mockConnectionPool.Verify(p => p.Clear(), Times.Once);
+        }
+
+        [Theory]
+        [InlineData(nameof(MongoConnectionException), true)]
+        [InlineData("MongoConnectionExceptionWithSocketTimeout", false)]
+        public void HandleChannelException_should_update_topology_as_expected_on_network_error_or_timeout(
+            string errorType, bool shouldUpdateTopology)
+        {
+            var serverId = new ServerId(_clusterId, _endPoint);
```

Comment:
it's a quite difficult test, do we really need to have it? Looking at the test name, we already have such functionality here: https://github.com/vincentkam/mongo-csharp-driver/pull/114/files#diff-5e394fc7c38046a2031b343c9f192d9cR552, isn't it?

Reply:
The setup is unfortunately complicated :( 

This test directly addresses the spec update requirements: that timeout exceptions do NOT update the topology and that network exceptions DO verify the topology, so I think this test is worth it. The other test indirectly addresses this requirement as well, and having both these tests does no harm I think, with this test **directly** testing the spec requirement and the updated test _indirectly_ testing it.

Reply:
Ok